### PR TITLE
PSTH

### PR DESCRIPTION
**Pull request by tensionhead**
## PSTH :rocket: 

- added peri stimulus time histogram with a frontend API like in [FT](https://github.com/fieldtrip/fieldtrip/blob/master/contrib/spike/ft_spike_psth.m)
- one minimal functional test was included, general proper tests still missing
- additional datasets like `avg` or `var` will get added once `TimeLockData` is up to speed

addresses #295 

Author Guidelines
-----------------
- [x] Was the code checked for memory leaks/performance bottlenecks?
- [x] Is the code running locally **and** on the ESI cluster?
- [x] Is the code running on all supported platforms?

Reviewer Checklist
------------------
- [x] Are testing routines present?
- [x] Do **parallel** loops have a set length and correct termination conditions?
- [x] Do objects in the global package namespace perform proper parsing of their input? 
- [x] Do code-blocks provide novel functionality, i.e., no re-factoring using builtin/external packages possible?
- [x] Code layout
  - [x] Is the code PEP8 compliant?
  - [x] Does the code adhere to agreed-upon naming conventions?
  - [x] Are keywords clearly named and easy to understand?
  - [x] No commented-out code?
- [ ] Are all docstrings complete and accurate?
- [ ] Is the CHANGELOG.md up to date?

### CODE REVIEW

File: syncopy/spikes/spike_psth.py [hunk 1]
```diff
@@ -84,21 +84,45 @@ def spike_psth(data,
...
+        sinfo = data.selection.trialdefinition[:, :2]
+        # why is this not directly data.selection.trials?!
+        trials = Indexer(map(data._get_trial, data.selection.trials),
+                         sinfo.shape[0])
+    else:
+        trl_def = data.trialdefinition
+        sinfo = data.sampleinfo
+        trials = data.trials
+
+    lenTrials = np.diff(sinfo).squeeze()
```
Comment (dfsp-spirit) [inline syncopy/spikes/spike_psth.py:107]: this seems to be unused

File: syncopy/spikes/spike_psth.py [hunk 2]
```diff
@@ -128,40 +152,64 @@ def spike_psth(data,
...
-
-    combs = np.unique(np.concatenate(combs), axis=0)
+    # it's a sequential loop to get an array of [chan, unit] indices
+    combs = get_chan_unit_combs(trials)
 
     # right away create the output labels for the channel axis
     chan_labels = [f'channel{i}_unit{j}' for i, j in combs]
 
     # now we have our global (single-trial, avg, std,..) histogram shape
     h_shape = (nBins, len(combs))
```
Comment (dfsp-spirit) [inline syncopy/spikes/spike_psth.py:179]: this is not part of this change set, but `chan_labels` and `h_shape` are also unused

**Comment by dfsp-spirit** [comment]
The docstring for the frontend, `spike_psth()`, lists a boolean parameter `vartriallen` that does not exist. Do we want to implement this later? Otherwise it should be removed.

File: syncopy/spikes/spike_psth.py (hunk 1 above)
Comment (tensionhead) [reply syncopy/spikes/spike_psth.py:107]: True, this I will remove!

File: syncopy/spikes/spike_psth.py [hunk 3]
```diff
@@ -84,21 +84,45 @@ def spike_psth(data,
...
     # --- parse and digest `latency` (time window of analysis) ---
 
+    # beginnings and ends of all trials in relative time
+    beg_ends = sinfo + trl_def[:, 2][:, None]
+    beg_ends = (beg_ends - sinfo[:, 0][:, None]) / data.samplerate
+
+    trl_starts = beg_ends[:, 0]
+    trl_ends = beg_ends[:, 1]
+    # just for sanity checks atm
+    tmin, tmax = trl_starts.min(), trl_ends.max()
```
[the remark this answers is not in the document]
Comment (tensionhead) [reply syncopy/spikes/spike_psth.py:118]: Yep, that's why the comment above.. might be that these guys are still gonna get used in the future, but I can just comment them in.

**Comment by dfsp-spirit** [comment]
The `output` parameter for the frontend, `spike_psth()`, seems to be ignored.

File: syncopy/spikes/psth.py [hunk 4]
```diff
@@ -45,41 +49,49 @@ def psth(trl_dat,
...
-    map_cu = lambda c, u: np.all(cu_combos == [c, u], axis=1)
+    map_cu = lambda c, u: np.all(chan_unit_combs == [c, u], axis=1)
+
     # now map with respect to unit for all single trial channels (-bins)
     map_unit = {u: np.logical_or.reduce([map_cu(c, u) for c in chan_vec]) for u in unique_units}
 
+    # map into histogram time x channel dimensions
+    map_unit_hist = {u: [np.any(map_cu(c, u)) for c in chan_vec] for u in unique_units}
+
     for i, iunit in enumerate(unique_units):
```
Comment (dfsp-spirit) [inline syncopy/spikes/psth.py:95]: again not part of this changeset, but `i` is unused, so no need for `enumerate`

File: syncopy/spikes/spike_psth.py [hunk 5]
```diff
@@ -7,30 +7,30 @@
 
 # Syncopy imports
 from syncopy.shared.parsers import data_parser, scalar_parser, array_parser
-from syncopy.shared.tools import get_defaults
-from syncopy.datatype import SpikeData
+from syncopy.shared.tools import get_defaults, get_frontend_cfg
+from syncopy.datatype import TimeLockData
+from syncopy.datatype.base_data import Indexer
 
 from syncopy.shared.errors import SPYValueError, SPYTypeError, SPYWarning, SPYInfo
```
Comment (dfsp-spirit) [inline syncopy/spikes/spike_psth.py:14]: minor: SPYTypeError, SPYWarning, SPYInfo are imported but not used

File: syncopy/spikes/psth.py [hunk 6]
```diff
@@ -3,10 +3,11 @@
...
-         cu_combos,
-         trl_start=0,
-         onset=0,
-         bins=None,
+         trl_start,
+         onset,
+         trl_end,
+         chan_unit_combs=None,
+         tbins=None,
          samplerate=1000):
```
Comment (dfsp-spirit) [inline syncopy/spikes/psth.py:11]: not part of this changeset, but the docstring for the `samplerate` parameter is missing

**Comment by tensionhead** [comment]
> The docstring for the frontend, `spike_psth()`, lists a boolean parameter `vartriallen` that does not exist. Do we want to implement this later? Otherwise it should be removed.

I implemented it just now :)

> The `output` parameter for the frontend, `spike_psth()`, seems to be ignored.

yep this still needs to be done!
